### DISCUSSION TITLE
DS18B20 ID SetOption for MQTT JSON

### DISCUSSION THREAD

**Discussion by pmknowles** [arendst/Tasmota#15509] [Ideas]
If a device with multiple DS18B20s has a faulty sensor, on a reboot, the 'internal' Tasmota sensor IDs re-order. If the third sensor (of, say, 5) becomes unreachable sensors 1 and 2 remain unchanged but 4 and 5 become 3 and 4. This is sensible because, otherwise, you would never be able to change the number of sensors.
However, this causes an issue for parsing the MQTT JSON string as
"DS18B20-4":{"Id":"0517C143A6FF","Temperature":18.3},"TempUnit":"C"}
would become
"DS18B20-3":{"Id":"0517C143A6FF","Temperature":18.3},"TempUnit":"C"}
and a lot of softwares will struggle to parse the change.
If there were a SetOption where 0 is to use the Tasmota 'internal' (DS18B20-1) ID and 1 is to 'drop' the internal ID just giving
{"DS18B20-Id":"0517C143A6FF","Temperature":18.3,"TempUnit":"C"}
it would make things simpler.
Another option would be to have the equivalent to WebButton and to be able to rename the sensor linked to the DS18B20 internal ID. This would help with oth the UI and MQTT.

**Comment by sfromis** [comment]
As [documented](https://tasmota.github.io/docs/DS18x20/#compile-options), you already have such an option, at build time, instead of a SetOption.
`#define DS18x20_USE_ID_AS_NAME      // Use last 3 bytes for naming of sensors`

**Reply by pmknowles** [reply]
Yes but do you think that is an adequate response?
In the 'old' days when we had to use Tasmota sensors should really have had that setting as the default
You have to compile your own build (which not eeryone would want to do)
The Tasmota internal ID is meaningless and as the numbering order is based on the hardware ID anyway why not just use the hardware ID - the ability to change WebButton makes the inability to change the sensor name look even sillier.
In theory it's possible to have 2 devices with the same last 3 bytes - yes you'd be unlucky but it could happen
Just using the hardware ID would also shorten the MQTT JSON reducing truncation (as the docs say MQTT buffer length is limited and SENSOR message will be truncated above 18 DS18x20. - Even less if other sensors are attached to the ESP device and present in the SENSOR message) .

**Reply by sfromis** [reply]
I merely added context about what's available, without an assumption of no improvement being useful, not trying to make a "case closed".

**Reply by barbudor** [reply]
You are right that in theory using only the last digit could lead to a dupplicate. Using the full ID could be better.
But the right JSON syntax would be:
```
"DS18B20-0517C143A6FF": { "Temperature":18.3}
```
as indeed repeating the ID is useless
However it may breaks existing integration as DS18x20 is used quite a lot
